// component-resolver/build/resolver.js

(function(
  // Reliable reference to the global object (i.e. window in browsers).
  global,

  // Dummy constructor that we use as the .constructor property for
  // functions that return Generator objects.
  GeneratorFunction,

  // Undefined value, more compressible than void 0.
  undefined
) {
  var hasOwn = Object.prototype.hasOwnProperty;

  if (global.wrapGenerator) {
    return;
  }

  function wrapGenerator(innerFn, self, tryList) {
    return new Generator(innerFn, self || null, tryList || []);
  }

  global.wrapGenerator = wrapGenerator;
  if (typeof exports !== "undefined") {
    exports.wrapGenerator = wrapGenerator;
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  wrapGenerator.mark = function(genFun) {
    genFun.constructor = GeneratorFunction;
    return genFun;
  };

  // Ensure isGeneratorFunction works when Function#name not supported.
  if (GeneratorFunction.name !== "GeneratorFunction") {
    GeneratorFunction.name = "GeneratorFunction";
  }

  wrapGenerator.isGeneratorFunction = function(genFun) {
    var ctor = genFun && genFun.constructor;
    return ctor ? GeneratorFunction.name === ctor.name : false;
  };

  function Generator(innerFn, self, tryList) {
    var generator = this;
    var context = new Context(tryList);
    var state = GenStateSuspendedStart;

    function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        throw new Error("Generator has already finished");
      }

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          try {
            var info = delegate.generator[method](arg);

            // Delegate generator ran and handled its own exceptions so
            // regardless of what the method was, we continue as if it is
            // "next" with an undefined arg.
            method = "next";
            arg = undefined;

          } catch (uncaught) {
            context.delegate = null;

            // Like returning generator.throw(uncaught), but without the
            // overhead of an extra function call.
            method = "throw";
            arg = uncaught;

            continue;
          }

          if (info.done) {
            context[delegate.resultName] = info.value;
            context.next = delegate.nextLoc;
          } else {
            state = GenStateSuspendedYield;
            return info;
          }

          context.delegate = null;
        }

        if (method === "next") {
          if (state === GenStateSuspendedStart &&
              typeof arg !== "undefined") {
            // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
            throw new TypeError(
              "attempt to send " + JSON.stringify(arg) + " to newborn generator"
            );
          }

          if (state === GenStateSuspendedYield) {
            context.sent = arg;
          } else {
            delete context.sent;
          }

        } else if (method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw arg;
          }

          if (context.dispatchException(arg)) {
            // If the dispatched exception was caught by a catch block,
            // then let that catch block handle the exception normally.
            method = "next";
            arg = undefined;
          }
        }

        state = GenStateExecuting;

        try {
          var value = innerFn.call(self, context);

          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          var info = {
            value: value,
            done: context.done
          };

          if (value === ContinueSentinel) {
            if (context.delegate && method === "next") {
              // Deliberately forget the last sent value so that we don't
              // accidentally pass it on to the delegate.
              arg = undefined;
            }
          } else {
            return info;
          }

        } catch (thrown) {
          state = GenStateCompleted;

          if (method === "next") {
            context.dispatchException(thrown);
          } else {
            arg = thrown;
          }
        }
      }
    }

    generator.next = invoke.bind(generator, "next");
    generator.throw = invoke.bind(generator, "throw");
  }

  Generator.prototype.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(triple) {
    var entry = { tryLoc: triple[0] };

    if (1 in triple) {
      entry.catchLoc = triple[1];
    }

    if (2 in triple) {
      entry.finallyLoc = triple[2];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry, i) {
    var record = entry.completion || {};
    record.type = i === 0 ? "normal" : "return";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryList.forEach(pushTryEntry, this);
    this.reset();
  }

  Context.prototype = {
    constructor: Context,

    reset: function() {
      this.prev = 0;
      this.next = 0;
      this.sent = undefined;
      this.done = false;
      this.delegate = null;

      this.tryEntries.forEach(resetTryEntry);

      // Pre-initialize at least 20 temporary variables to enable hidden
      // class optimizations for simple generators.
      for (var tempIndex = 0, tempName;
           hasOwn.call(this, tempName = "t" + tempIndex) || tempIndex < 20;
           ++tempIndex) {
        this[tempName] = null;
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    keys: function(object) {
      var keys = [];
      for (var key in object) {
        keys.push(key);
      }
      keys.reverse();

      // Rather than returning an object with a next method, we keep
      // things simple and return the next function itself.
      return function next() {
        while (keys.length) {
          var key = keys.pop();
          if (key in object) {
            next.value = key;
            next.done = false;
            return next;
          }
        }

        // To avoid creating an additional object, we just hang the .value
        // and .done properties off the next function object itself. This
        // also ensures that the minifier will not anonymize the function.
        next.done = true;
        return next;
      };
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;
        return !!caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    _findFinallyEntry: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") && (
              entry.finallyLoc === finallyLoc ||
              this.prev < entry.finallyLoc)) {
          return entry;
        }
      }
    },

    abrupt: function(type, arg) {
      var entry = this._findFinallyEntry();
      var record = entry ? entry.completion : {};

      record.type = type;
      record.arg = arg;

      if (entry) {
        this.next = entry.finallyLoc;
      } else {
        this.complete(record);
      }

      return ContinueSentinel;
    },

    complete: function(record) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = record.arg;
        this.next = "end";
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      var entry = this._findFinallyEntry(finallyLoc);
      return this.complete(entry.completion);
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry, i);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(generator, resultName, nextLoc) {
      this.delegate = {
        generator: generator,
        resultName: resultName,
        nextLoc: nextLoc
      };

      return ContinueSentinel;
    }
  };
}).apply(this, Function("return [this, function GeneratorFunction(){}]")());

var co = require('co');
var chanel = require('chanel');
var remotes = require('remotes');
var resolve = require('path').resolve;
var inherits = require('util').inherits;
var EventEmitter = require('events').EventEmitter;
var downloader = require('component-downloader');
var debug = require('debug')('component-resolver');
module.exports = Resolver;
inherits(Resolver, EventEmitter);

/**
 * Resolve a component's locals and dependencies.
 * `root` can either be the component.json itself or the working directory.
 *
 * @param {String|Object} root || component
 * @param {Object} options
 * @api public
 */

function Resolver(root, options) {
  options = options || {};
  EventEmitter.call(this, options)

  var concurrency = options.concurrency || {};

  if (!root || typeof root === 'string') {
    this.root = root || process.cwd();
  } else if (typeof root === 'object') {
    // all components need a name!
    // we fix objects as arguments so you can just
    // do {dependencies: {}}
    root.name = root.name || 'root';
    this.main = root;
    this.root = options.root || process.cwd();
  } else {
    throw new TypeError('wtf');
  }

  // include development stuff, false by default
  this.dev = options.dev || options.development;
  // resolve dependencies, true by default
  this.deps = options.dependencies !== false;
  // output folder for installs
  this.out = resolve(options.out
    || options.dir
    || 'components');
  // silent by default
  this.verbose = !!options.verbose;
  // do not log validation warnings by default
  // as it would just make things noisey.
  this.validate = !!options.validate;

  // if you want to set your own remote options
  // like auth, netrc, proxy, etc.,
  // then create your own remote instance.
  if (!(this.remote = options.remote || options.remotes)) {
    this.remote = remotes(remotes.defaults, {
      out: this.out,
      // we check local components by default
      local: options.local !== false,
      timeout: options.timeout || null,
      retries: options.retries || null
    });
    debug('remote not set - defaulting to remotes\'s defaults');
  }

  this.install = !!options.install;
  if (this.install) {
    // fields to check to download files
    this._download = downloader({
      out: this.out,
      remote: this.remote,
      fields: options.fields,
      concurrency: concurrency.downloads || 1,
      verbose: this.verbose,
      archive: options.archive,
    })
  }

  this.channel = {
    locals: chanel({
      // local fs calls.
      // concurrency isn't really necessary here
      // since graceful-fs should handle any EMFILE errors
      concurrency: concurrency.locals || 16,
      discard: true
    }),
    dependencies: chanel({
      // concurrent GET component.json
      concurrency: concurrency.dependencies || 5,
      discard: true
    }),
    semver: chanel({
      // concurrent GET versions calls
      // hitting issues when there are more than a couple
      // channels for semver resolution
      concurrency: concurrency.semver || 1,
      discard: true
    }),
    installs: chanel({
      // concurrent installations
      concurrency: concurrency.installs || 5,
      discard: true
    }),
  }

  // keep track of branches in a flat list
  // so you don't have to traverse a tree
  // has no order!
  this.locals = [];
  this.dependencies = [];

  // keep track of resolution states
  this.state = {};
  // keep track of local component.jsons
  this.cache = {};
  // keep track of any hangs
  this.status = {};
}

/**
 * Call this to execute the resolver and return the tree.
 *
 * @return {Object} tree
 * @api public
 */

Resolver.prototype.tree = wrapGenerator.mark(function() {
  var tree, ch;

  return wrapGenerator(function($ctx0) {
    while (1) switch ($ctx0.prev = $ctx0.next) {
    case 0:
      return $ctx0.delegateYield(this.branchLocal(this.root, this.main), "t0", 1);
    case 1:
      tree = this.branch = $ctx0.t0;

      // local and github remotes by default
      // if `local: false`, then the `local` remote will be skipped
      if (!tree.remotes.length) tree = tree.concat(remotes.defaults);

      ch = this.channel;
      $ctx0.next = 6;
      return ch.locals(true);
    case 6:
      debug('finished resolving locals');
      $ctx0.next = 9;
      return ch.dependencies(true);
    case 9:
      debug('finished resolving dependencies (1)');

      // start resolving all the hanging semvers after
      // all the locals and dependencies are done.
      // not ideal as we should do this ASAP,
      // but i'm not sure how.
      this.semver();
    case 11:
      if (!(ch.semver.readable || ch.dependencies.readable)) {
        $ctx0.next = 20;
        break;
      }

      $ctx0.next = 14;
      return ch.semver(true);
    case 14:
      debug('finished resolving semver');
      $ctx0.next = 17;
      return ch.dependencies(true);
    case 17:
      debug('finished resolving dependencies(2)');
      $ctx0.next = 11;
      break;
    case 20:
      $ctx0.next = 22;
      return ch.installs(true);
    case 22:
      debug('finished installing dependencies');
      return $ctx0.abrupt("return", tree);
    case 24:
    case "end":
      return $ctx0.stop();
    }
  }, this);
})

/**
 * Vanilla JS method of `tree = yield* resolver.tree()`.
 *
 * @api public
 */;

/**
 * Vanilla JS method of `tree = yield* resolver.tree()`.
 *
 * @api public
 */

Resolver.prototype.getTree = co(Resolver.prototype.tree);

/**
 * Resolve a branch's dependency and local branches.
 *
 * @param {Object} branch
 * @api private
 */

Resolver.prototype.resolve = function (branch) {
  debug('resolving "%s"', branch.name);
  var component = branch.node
  var development = this.dev
    && branch.type === 'local'
    && component.development

  // resolve dependencies in a different channel
  if (this.deps) {
    this.resolveDependencies(branch, component.dependencies);
    if (development) this.resolveDependencies(branch, development.dependencies);
  }

  // resolve locals in a different channel
  this.resolveLocals(branch, component.locals);
  if (development) this.resolveLocals(branch, development.locals);

  debug('remaining dependencies: ' + this.channel.dependencies.queue);
  debug('remaining semver: ' + this.channel.semver.queue);
}

/**
 * Await an event.
 *
 * @param {String} event
 * @returns {Object} branch
 * @api private
 */;

/**
 * Await an event.
 *
 * @param {String} event
 * @returns {Object} branch
 * @api private
 */

Resolver.prototype.await = function (event) {
  var self = this
  return function (done) {
    self.once(event, function (branch) {
      done(null, branch);
    });
  };
};

/**
 * Get all the remotes to use in a branch's dependencies.
 *
 * @param {Object} branch
 * @return {Array} remotes
 * @api private
 */

Resolver.prototype.resolveRemotes = function (branch) {
  var remotes = branch.resolvedRemotes = [];
  do {
    var names = branch.remotes || [];
    for (var i = 0; i < names.length; i++)
      if (!~remotes.indexOf(names[i]))
        remotes.push(names[i]);
  } while (branch = branch.parent)
  return remotes;
}
