// cogent/build/index.js
(function(
  // Reliable reference to the global object (i.e. window in browsers).
  global,

  // Dummy constructor that we use as the .constructor property for
  // functions that return Generator objects.
  GeneratorFunction,

  // Undefined value, more compressible than void 0.
  undefined
) {
  var hasOwn = Object.prototype.hasOwnProperty;

  if (global.wrapGenerator) {
    return;
  }

  function wrapGenerator(innerFn, self, tryList) {
    return new Generator(innerFn, self || null, tryList || []);
  }

  global.wrapGenerator = wrapGenerator;
  if (typeof exports !== "undefined") {
    exports.wrapGenerator = wrapGenerator;
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  wrapGenerator.mark = function(genFun) {
    genFun.constructor = GeneratorFunction;
    return genFun;
  };

  // Ensure isGeneratorFunction works when Function#name not supported.
  if (GeneratorFunction.name !== "GeneratorFunction") {
    GeneratorFunction.name = "GeneratorFunction";
  }

  wrapGenerator.isGeneratorFunction = function(genFun) {
    var ctor = genFun && genFun.constructor;
    return ctor ? GeneratorFunction.name === ctor.name : false;
  };

  function Generator(innerFn, self, tryList) {
    var generator = this;
    var context = new Context(tryList);
    var state = GenStateSuspendedStart;

    function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        throw new Error("Generator has already finished");
      }

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          try {
            var info = delegate.generator[method](arg);

            // Delegate generator ran and handled its own exceptions so
            // regardless of what the method was, we continue as if it is
            // "next" with an undefined arg.
            method = "next";
            arg = undefined;

          } catch (uncaught) {
            context.delegate = null;

            // Like returning generator.throw(uncaught), but without the
            // overhead of an extra function call.
            method = "throw";
            arg = uncaught;

            continue;
          }

          if (info.done) {
            context[delegate.resultName] = info.value;
            context.next = delegate.nextLoc;
          } else {
            state = GenStateSuspendedYield;
            return info;
          }

          context.delegate = null;
        }

        if (method === "next") {
          if (state === GenStateSuspendedStart &&
              typeof arg !== "undefined") {
            // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
            throw new TypeError(
              "attempt to send " + JSON.stringify(arg) + " to newborn generator"
            );
          }

          if (state === GenStateSuspendedYield) {
            context.sent = arg;
          } else {
            delete context.sent;
          }

        } else if (method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw arg;
          }

          if (context.dispatchException(arg)) {
            // If the dispatched exception was caught by a catch block,
            // then let that catch block handle the exception normally.
            method = "next";
            arg = undefined;
          }
        }

        state = GenStateExecuting;

        try {
          var value = innerFn.call(self, context);

          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          var info = {
            value: value,
            done: context.done
          };

          if (value === ContinueSentinel) {
            if (context.delegate && method === "next") {
              // Deliberately forget the last sent value so that we don't
              // accidentally pass it on to the delegate.
              arg = undefined;
            }
          } else {
            return info;
          }

        } catch (thrown) {
          state = GenStateCompleted;

          if (method === "next") {
            context.dispatchException(thrown);
          } else {
            arg = thrown;
          }
        }
      }
    }

    generator.next = invoke.bind(generator, "next");
    generator.throw = invoke.bind(generator, "throw");
  }

  Generator.prototype.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(triple) {
    var entry = { tryLoc: triple[0] };

    if (1 in triple) {
      entry.catchLoc = triple[1];
    }

    if (2 in triple) {
      entry.finallyLoc = triple[2];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry, i) {
    var record = entry.completion || {};
    record.type = i === 0 ? "normal" : "return";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryList.forEach(pushTryEntry, this);
    this.reset();
  }

  Context.prototype = {
    constructor: Context,

    reset: function() {
      this.prev = 0;
      this.next = 0;
      this.sent = undefined;
      this.done = false;
      this.delegate = null;

      this.tryEntries.forEach(resetTryEntry);

      // Pre-initialize at least 20 temporary variables to enable hidden
      // class optimizations for simple generators.
      for (var tempIndex = 0, tempName;
           hasOwn.call(this, tempName = "t" + tempIndex) || tempIndex < 20;
           ++tempIndex) {
        this[tempName] = null;
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    keys: function(object) {
      var keys = [];
      for (var key in object) {
        keys.push(key);
      }
      keys.reverse();

      // Rather than returning an object with a next method, we keep
      // things simple and return the next function itself.
      return function next() {
        while (keys.length) {
          var key = keys.pop();
          if (key in object) {
            next.value = key;
            next.done = false;
            return next;
          }
        }

        // To avoid creating an additional object, we just hang the .value
        // and .done properties off the next function object itself. This
        // also ensures that the minifier will not anonymize the function.
        next.done = true;
        return next;
      };
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;
        return !!caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    _findFinallyEntry: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") && (
              entry.finallyLoc === finallyLoc ||
              this.prev < entry.finallyLoc)) {
          return entry;
        }
      }
    },

    abrupt: function(type, arg) {
      var entry = this._findFinallyEntry();
      var record = entry ? entry.completion : {};

      record.type = type;
      record.arg = arg;

      if (entry) {
        this.next = entry.finallyLoc;
      } else {
        this.complete(record);
      }

      return ContinueSentinel;
    },

    complete: function(record) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = record.arg;
        this.next = "end";
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      var entry = this._findFinallyEntry(finallyLoc);
      return this.complete(entry.completion);
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry, i);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(generator, resultName, nextLoc) {
      this.delegate = {
        generator: generator,
        resultName: resultName,
        nextLoc: nextLoc
      };

      return ContinueSentinel;
    }
  };
}).apply(this, Function("return [this, function GeneratorFunction(){}]")());

var url = require('url');
var util = require('util');
var zlib = require('zlib');
var http = require('http');
var https = require('https');
var Netrc = require('netrc');
var write = require('write-to');
var status = require('statuses');
var rawBody = require('raw-body');
var proxyAgent = require('proxy-agent');
var debug = require('debug')('cogent');
var inspect = util.inspect;
exports = module.exports = extend();
exports.extend = extend;

function extend(defaults) {
  wrapGenerator.mark(cogent);
  defaults = defaults || {}
  defaults.retries = defaults.retries || 0;
  defaults.redirects = 'redirects' in defaults ? defaults.redirects : 3;
  defaults.timeout = defaults.timeout || 5000;
  defaults.method = defaults.method || 'GET';
  defaults.gunzip = defaults.gunzip === false ? false : true;
  defaults.netrc = Netrc(defaults.netrc);
  return cogent;

  function cogent(uri, options) {
    var redirects, timeout, netrc, headers, urls, o, req, res, code, stream, securrrr, retries, auth, agent, gunzip, request;

    return wrapGenerator(function cogent$($ctx0) {
      while (1) switch ($ctx0.prev = $ctx0.next) {
      case 0:
        request = function request(done) {
          var called = false

          // timeout
          // note: timeout is only for the response,
          // not the entire request
          var id = setTimeout(function () {
            debug('timeout exceeded for %s %s', o.method, o.href)
            var err = new Error('timeout of ' + timeout + 'ms exceeded for "' + o.href + '"')
            err.url = o
            next(err)
          }, timeout)

          req = (securrrr ? https : http).request(o)
          req.once('response', next.bind(null, null))
          req.once('error', next)
          req.end()

          function next(err, res) {
            if (called) {
              // dump the stream in case there are any
              // to avoid any possible leaks
              if (res) res.resume()
              return
            }

            called = true
            clearTimeout(id)

            // kill the request, specifically for timeouts
            // to do: tests for this #7
            if (err) req.abort()

            if (retries && (err
              || (res && status.retry[res.statusCode])
            )) {
              debug('retrying %s %s', o.method, o.href)
              retries--
              request(done)
            } else if (err) {
              debug('received error "%s" with "%s"', err.message, o.href)
              done(err)
            } else {
              done(null, res)
            }
          }
        };

        // options type checking stuff
        if (options === true)
          options = { json: true }
        else if (typeof options === 'string')
          options = { destination: options }
        else if (!options)
          options = {}

        // setup defaults;

        redirects = options.redirects || defaults.redirects;
        timeout = options.timeout || defaults.timeout;
        netrc = options.netrc ? Netrc(options.netrc) : defaults.netrc;
        headers = options.headers = options.headers || {};
        headers['accept-encoding'] = headers['accept-encoding'] || 'gzip';
        headers['user-agent'] = headers['user-agent'] || 'https://github.com/cojs/cogent';

        if (options.json) headers.accept = 'application/json'

        // keep a history of redirect urls;

        urls = [];
      case 10:
        if (!true) {
          $ctx0.next = 60;
          break;
        }

        // create the request options object
        urls.push(o = url.parse(uri));

        securrrr = o.protocol === 'https:';
        o.headers = options.headers;

        o.method = options.method || defaults.method

        // auth, either by .auth or by .netrc;

        // auth, either by .auth or by .netrc
        if (options.auth || defaults.auth) {
          o.auth = options.auth || defaults.auth
        } else {
          auth = netrc[o.hostname];

          if (auth) o.auth =
              (auth.login || auth.user || auth.username)
            + ':'
            + (auth.pass || auth.password);
        }

        // setup agent or proxy agent
        if ('agent' in options) {
          o.agent = options.agent
        } else if (options.proxy || defaults.proxy) {
          agent = proxyAgent(options.proxy || defaults.proxy, securrrr);

          if (agent)
            o.agent = agent;
        } else if (defaults.agent != null) {
          o.agent = defaults.agent
        }

        // retries is on a per-URL-request basis
        retries = options.retries || defaults.retries;

        debug('options: %s', inspect(o));
        $ctx0.next = 21;
        return request;
      case 21:
        res = $ctx0.sent;

        code = res.statusCode

        // redirect;

        if (!(redirects-- && status.redirect[code])) {
          $ctx0.next = 28;
          break;
        }

        debug('redirecting from %s to %s', uri, res.headers.location);
        uri = url.resolve(uri, res.headers.location);

        // dump this stream
        res.resume() // dump this stream;

        return $ctx0.abrupt("continue", 10);
      case 28:
        if (!(o.method.toUpperCase() === 'HEAD')) {
          $ctx0.next = 32;
          break;
        }

        res.req = req;
        res.resume();
        return $ctx0.abrupt("return", res);
      case 32:
        gunzip = typeof options.gunzip === 'boolean'
          ? options.gunzip
          : defaults.gunzip;

        if (!(!gunzip && (options.buffer || options.string || options.json))) {
          $ctx0.next = 35;
          break;
        }

        throw new Error('must gunzip if buffering the response');
      case 35:
        if (res.headers['content-encoding'] === 'gzip' && gunzip) {
          stream = res.pipe(zlib.createGunzip())
          stream.res = res
          // pass useful response stuff
          stream.statusCode = code
          stream.headers = res.headers
          // forward errors
          res.on('error', stream.emit.bind(stream, 'error'))
          // forward close
          // creates the following error:
          // Assertion failed: (!write_in_progress_ && "write in progress"), function Close, file ../src/node_zlib.cc, line 99.
          // make: *** [test] Abort trap: 6
          // res.on('close', stream.close.bind(stream))
          // forward destroy
          // note: if zlib adds a .destroy method, we have to change this
          stream.destroy = res.destroy.bind(res)
          res = stream
        } else {
          res.res = res
        }

        res.req = req;

        res.urls = urls

        // save the file;

        if (!(code === 200 && options.destination)) {
          $ctx0.next = 43;
          break;
        }

        $ctx0.next = 41;
        return write(res, options.destination);
      case 41:
        res.destination = options.destination;
        return $ctx0.abrupt("return", res);
      case 43:
        if (!options.buffer) {
          $ctx0.next = 47;
          break;
        }

        $ctx0.next = 46;
        return rawBody(res);
      case 46:
        res.buffer = $ctx0.sent;
      case 47:
        if (!(options.string || options.json)) {
          $ctx0.next = 56;
          break;
        }

        if (!res.buffer) {
          $ctx0.next = 52;
          break;
        }

        $ctx0.t0 = res.buffer.toString('utf8');
        $ctx0.next = 55;
        break;
      case 52:
        $ctx0.next = 54;
        return rawBody(res, { encoding: options.string || true });
      case 54:
        $ctx0.t0 = $ctx0.sent;
      case 55:
        res.text = $ctx0.t0;
      case 56:
        // buffer the response as JSON
        if (options.json) try {
          res.body = JSON.parse(res.text)
        } catch (err) {
          debug('"%s" is not valid JSON', uri)
        }

        return $ctx0.abrupt("return", res);
      case 60:
      case "end":
        return $ctx0.stop();
      }
    }, this);
  }
}
